amaranth/sim/_pycoro.py

```diff
@@ -58,29 +59,38 @@ def run(self):
         if self.coroutine is None:
             return
 
-        self.clear_triggers()
+        if self.command is None:
+            self.clear_triggers()
 
         response = None
         exception = None
+        assigned = False
         while True:
-            try:
-                if exception is None:
-                    command = self.coroutine.send(response)
-                else:
-                    command = self.coroutine.throw(exception)
-            except StopIteration:
-                self.passive = True
-                self.coroutine = None
-                return False # no assignment
+            if self.command is not None:
+                command, self.command = self.command, None
+            else:
+                try:
+                    if exception is None:
+                        command = self.coroutine.send(response)
+                    else:
+                        command = self.coroutine.throw(exception)
+                except StopIteration:
+                    self.passive = True
+                    self.coroutine = None
+                    return assigned
 
             try:
                 if command is None:
                     command = self.default_cmd
+                if isinstance(command, ValueCastable):
+                    command = Value.cast(command)
                 response = None
                 exception = None
 
-                if isinstance(command, ValueCastable):
-                    command = Value.cast(command)
+                if assigned and isinstance(command, (Value, Tick, Settle, Delay, MemorySimRead)):
```

Comment:
cursed: this also needs `Statement`, unless you can prove its `_rhs_signals` is empty — after all, assignments can also *read* signals

Reply:
same for `_MemorySimWrite` actually, its `_addr` and `_data` are currently allowed to be arbitrary expressions

Reply:
This... makes the second commit completely useless, so either a check on `_rhs_signals()` must be added, or the second commit should be rolled back.

Reply:
I took the 2nd commit out of the PR.